### PR TITLE
fix #36

### LINKED ISSUE #36: Improve input method support for windows [$20]
[Issuehunt badges](https://issuehunt.io/r/lc-soft/LCUI/issues/36)

**Problem**
- The top left of the screen will display an input box.
- The candidate box is positioned at the top left of the screen.

image

**Expected effect**
- No additional controls will be displayed.
- The candidate box should be positioned at the text caret, like this:

image

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $20.00)

- [lc-soft lc-soft](https://issuehunt.io/u/lc-soft) ($20.00)
### Submitted pull Requests
- [#175 fix #36](https://issuehunt.io/r/lc-soft/LCUI/pull/175)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/lc-soft/LCUI/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by 1354092549**
---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#36: Improve input method support for windows [$20]](https://issuehunt.io/repos/5293802/issues/36)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by lc-soft** [comment]
Can you upload screenshots ?

File: include/LCUI/main.h [hunk 1]
```diff
@@ -53,6 +53,7 @@ enum LCUI_SysEventType {
 	LCUI_PAINT,
 	LCUI_WIDGET,
 	LCUI_QUIT, /**< 在 LCUI 退出前触发的事件 */
+	LCUI_CARETPOSCHANGED,
```
Comment (lc-soft) [inline include/LCUI/main.h]: It is an input method related event, should be prefixed with IME, and CARETPOSCHANGED is not easy to read and should be changed to SETCARET, so you should rename it to `LCUI_IME_SETCARET`, just like `WM_SETCURSOR` and `WM_IME_SETCONTEXT`

File: src/platform/windows/windows_ime.c [hunk 2]
```diff
@@ -32,14 +32,14 @@
...
 #include <LCUI_Build.h>
-
 #ifdef LCUI_BUILD_IN_WIN32
 #include <LCUI/LCUI.h>
 #include <LCUI/input.h>
 #include <LCUI/ime.h>
 #include <LCUI/platform.h>
 #include LCUI_EVENTS_H
 
+#pragma comment(lib, "Imm32.lib")
```
Comment (lc-soft) [inline src/platform/windows/windows_ime.c]: Need to insert a new blank line.

File: src/platform/windows/windows_ime.c [hunk 3]
```diff
@@ -60,14 +60,39 @@ static void WinIME_OnChar(LCUI_Event e, void *arg)
...
+		ImmSetCompositionWindow(hIMC, &Composition);
+
+		ImmReleaseContext(hwnd, hIMC);
+	}
+}
+
+static int OnCaretPosChanged_Id;
 static LCUI_BOOL IME_Open(void)
 {
+	OnCaretPosChanged_Id = LCUI_BindEvent(
```
Comment (lc-soft) [inline src/platform/windows/windows_ime.c]: Variable names should be lowercase

File: src/platform/windows/windows_ime.c [hunk 4]
```diff
@@ -60,14 +60,39 @@ static void WinIME_OnChar(LCUI_Event e, void *arg)
...
+		Composition.dwStyle = CFS_POINT;
+		Composition.ptCurrentPos.x = point.x;
+		Composition.ptCurrentPos.y = point.y;
+		ImmSetCompositionWindow(hIMC, &Composition);
+
+		ImmReleaseContext(hwnd, hIMC);
+	}
+}
+
+static int OnCaretPosChanged_Id;
```
Comment (lc-soft) [inline src/platform/windows/windows_ime.c]: All global variables should be centrally defined at the top, you should move it to the front of IME_ProcessKey()

File: src/platform/windows/windows_ime.c [hunk 5]
```diff
@@ -60,14 +60,39 @@ static void WinIME_OnChar(LCUI_Event e, void *arg)
...
+static void WinIME_OnCaretPosChanged(LCUI_SysEvent e, void *arg)
+{
+	HWND hwnd = GetActiveWindow();
+	HIMC hIMC = ImmGetContext(hwnd);
+	LCUI_Pos caret_pos;
+	if (hIMC) {
+		POINT point;
+		caret_pos = LCUI_GetCaretPos();
+		point.x = caret_pos.x;
+		point.y = caret_pos.y;
```
Comment (lc-soft) [inline src/platform/windows/windows_ime.c]: Just copy the value of caret_pos to point ? Why not use caret_pos directly?

``` diff
- Composition.ptCurrentPos.x = point.x;
- Composition.ptCurrentPos.y = point.y;
+ Composition.ptCurrentPos.x = caret_pos.x;
+ Composition.ptCurrentPos.y = caret_pos.y;
```

**Comment by 1354092549** [comment]
I'm sorry that I don't know much about the code structure of this project.
I have pushed a new version to implement SetCaret by using a normal IME method.
Plus: Can I use Chinese here?

**Comment by lc-soft** [comment]
能用中文就用中文，整英文反而增加沟通成本。

File: src/ime.c [hunk 6]
```diff
@@ -270,6 +270,13 @@ void LCUI_InitIME(void)
 #endif
 }
 
+void LCUIIME_SetCaret(LCUI_Pos pos)
+{
+	if (self.ime->handler.setcaret == NULL)
```
Comment (lc-soft) [inline src/ime.c]: if 里就算只有一行语句也应该加花括号。

File: src/gui/widget/textcaret.c [hunk 7]
```diff
@@ -60,10 +61,16 @@ textcaret {
 void TextCaret_Refresh(LCUI_Widget widget)
 {
 	LCUI_TextCaret caret = Widget_GetData(widget, prototype);
+	LCUI_Pos pos;
+	float x, y;
```
Comment (lc-soft) [inline src/gui/widget/textcaret.c]: 变量定义块的末尾应该空一行。

Comment (lc-soft) [reply src/gui/widget/textcaret.c]: 最好是按宽度排序，基础类型和自定义类的变量定义也应该割开。

``` c
float x,y;

LCUI_Pos pos;
LCUI_TextCaret caret = Widget_GetData(widget, prototype);

if (caret->visible) {
        return;
}
```

File: src/ime.c (hunk 6 above)
Comment (lc-soft) [reply src/ime.c]: 只是处理 if 条件成立的情况，那可以写成：

``` c
if (self.ime->handler.setcaret) {
        self.ime->handler.setcaret(pos);
}
```

File: src/platform/windows/windows_ime.c [hunk 8]
```diff
@@ -60,6 +61,20 @@ static void WinIME_OnChar(LCUI_Event e, void *arg)
 	LCUIIME_Commit(text, 2);
 }
 
+static void IME_SetCaret(LCUI_Pos pos)
+{
+	HWND hwnd = GetActiveWindow();
+	HIMC hIMC = ImmGetContext(hwnd);
+	if (hIMC) {
+		COMPOSITIONFORM Composition;
```
Comment (lc-soft) [inline src/platform/windows/windows_ime.c]: Composition 改成小写 composition

File: src/ime.c [hunk 9]
```diff
@@ -270,6 +270,13 @@ void LCUI_InitIME(void)
 #endif
 }
 
+void LCUIIME_SetCaret(LCUI_Pos pos)
```
Comment (lc-soft) [inline src/ime.c]: LCUI_Pos 是以前定义数据类型，现在已经尽量不用它了，请使用 x  y 两个变量代替

File: src/platform/windows/windows_ime.c [hunk 10]
```diff
@@ -60,6 +61,20 @@ static void WinIME_OnChar(LCUI_Event e, void *arg)
 	LCUIIME_Commit(text, 2);
 }
 
+static void IME_SetCaret(LCUI_Pos pos)
+{
+	HWND hwnd = GetActiveWindow();
+	HIMC hIMC = ImmGetContext(hwnd);
```
Comment (lc-soft) [inline src/platform/windows/windows_ime.c]: hIMC 改成小写 himc

**Comment by 1354092549** [comment]
(PR updated)

**Comment by lc-soft** [comment]
@1354092549 效果图发一下，最好是 gif 屏幕录像，windows 系统可以用 [ScreenToGif](https://www.screentogif.com/?l=zh_cn) 录制。

**Comment by 1354092549** [comment]
GIF
另外考虑一下布置cpplint之类的检查器？

**Comment by lc-soft** [comment]
> 另外考虑一下布置cpplint之类的检查器？

现在用的是 clang-format

**Comment by 1354092549** [comment]
> 现在用的是 clang-format

clang-format貌似不包括 命名检查？

**Comment by lc-soft** [comment]
> clang-format貌似不包括 命名检查？

@1354092549 是的，只是个格式化工具，没检查功能。

**Pull request merged**